pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,338 @@
+	// Retrieve node from reconcile request
+	node := &v1.Node{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to node deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				c.deleteGaguges(labels)
+			} else {
+				logging.FromContext(ctx).Errorf("Failed to delete gauge: failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after node labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+		c.deleteGaguges(labels)
+	}
+	newlabels, err := c.generateLabels(node)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate labels: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	if err := c.updateGaguges(ctx, node, newlabels); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to update gauges: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+	c.LabelsMap[req.NamespacedName] = newlabels
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("nodemetrics").
+		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				nodes := &v1.NodeList{}
+				if err := c.KubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{v1alpha5.ProvisionerNameLabelKey: o.GetName()})); err != nil {
+					logging.FromContext(ctx).Errorf("Failed to list nodes when mapping expiration watch events, %s", err.Error())
+					return requests
+				}
+				for _, node := range nodes.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+				}
+				return requests
+			}),
+		).
+		Watches(
+			// Reconcile node when a pod assigned to it changes.
+			&source.Kind{Type: &v1.Pod{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				if name := o.(*v1.Pod).Spec.NodeName; name != "" {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: name}})
+				}
+				return requests
+			}),
+		).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10000}).
+		Complete(c)
+	return err
+}
+
+func (c *Controller) deleteGaguges(labels *prometheus.Labels) {
+	for _, gaugeVec := range c.GaugeVecMap {
+		gaugeVec.Delete(*labels)
+	}
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(node *v1.Node) (*prometheus.Labels, error) {
+	metricLabels := prometheus.Labels{}
+
+	metricLabels[nodeName] = node.GetName()
+	metricLabels[nodeProvisioner] = node.Labels[v1alpha5.ProvisionerNameLabelKey]
```

Comment:
Is it crazy to include *every* label on the node in a `for`?

Reply:
The labels need to be declared when creating the gaugeVec. It is anti-pattern to change the labels (not the values, but the definition) on the fly. So it is impossible to just add new prometheus labels from node's k8s labels without some serious hacking. As a result, I add all the node k8s labels as the value of a prometheus label called nodeLabel.

Reply:
Does this mean that it isn't possible for users to add their own label filters when looking at dashboard, or are these functions independent of each other?

Reply:
User can add prometheus label filters based on values. But they can't use prometheus label that are not defined in the controller.  Therefore, if user want to add a K8S label and want to filter according to that, we will have to use a workaround. The hack is to put all the K8S label key value pairs under one prometheus label (i called it pod_labels) as a json file. Then we do string parsing and filtering regarding to this.

After some discussion we conclude that there is no good use case at this time. Maybe we can revisit this idea later.

Reply:
I see. Can you cut an issue here for tracking in the future so that we don't forget about it? 

If possible, it would be good to have a non-hacky way to do this. When we have a better understanding of what metrics we'll need and what use-cases are needed, I think we could have another discussion.